File: backend/games/TicTacToe.py
import numpy as np

class TicTacToe:
    def __init__(self):
        self.board = np.zeros((3, 3))
        self.current_player = 1

    def reset(self):
        self.board = np.zeros((3, 3))
        self.current_player = 1
        return self.board.tolist()

    def move(self, position):
        if self.board[position[0]][position[1]] != 0:
            return False
        self.board[position[0]][position[1]] = self.current_player
        self.current_player = self.current_player * -1  # switch player
        return True

    def check_winner(self):
        print(self.board)
        for i in range(len(self.board)):
            if self.board[i][0] == self.board[i][1] == self.board[i][2] != 0:
                return self.board[i][0]
        for i in range(len(self.board[0])):
            if self.board[0][i] == self.board[1][i] == self.board[2][i] != 0:
                return self.board[0][i]
        if self.board[0][0] == self.board[1][1] == self.board[2][2] != 0:
            return self.board[0][0]
        if self.board[0][2] == self.board[1][1] == self.board[2][0] != 0:
            return self.board[0][2]
        if 0 not in self.board:
            return 0  # Draw
        return None  # No winner yet

File: frontend/src/games/ticTacToe/TicTacToe.js
// src/TicTacToe.js
import React, { useState, useEffect } from 'react';
import axios from 'axios';
import './TicTacToe.css';

const TicTacToe = () => {
    const [board, setBoard] = useState(Array(3).fill(Array(3).fill(null)));
    const [winner, setWinner] = useState(null);
    const [currentPlayer, setCurrentPlayer] = useState(1);
    const [status, setStatus] = useState('Next player: X');

    useEffect(() => {
        resetGame();
    }, []);

    // Base URL for the backend
    const BASE_URL = 'http://localhost:5000';

    const resetGame = async () => {
        try {
            const response = await axios.post(`${BASE_URL}/tictactoe-reset`);
            setBoard(response.data);
            setWinner(null);
            setCurrentPlayer(response.data.currentPlayer);
        } catch (error) {
            console.error('Error resetting the game:', error);
        }
    };

    const handleClick = async (i,j) => {
        if (board[i][j] || winner !== null) return;
        try {
            const response = await axios.post(`${BASE_URL}/tictactoe-move`, { position: [i,j] });
            console.log(response.data);
            setBoard(response.data.board);
            setWinner(response.data.winner);
            setCurrentPlayer((prevPlayer) => (prevPlayer === 1 ? -1 : 1));
        } catch (error) {
            console.error('Invalid move:', error);
        }
    };

    useEffect(() => {
        changestatus();
    }, [winner, currentPlayer]);

    const renderSquare = (i,j) => (
        <button className="square" onClick={() => handleClick(i,j)}>
            {board[i][j] === 1 ? 'X' : board[i][j] === -1 ? 'O' : ''}
        </button>
    );

    function changestatus () {
        if (winner === null) {
            if (currentPlayer === 1) {
                setStatus('Next player: O');
            } else {
                setStatus('Next player: X');
            }
        } else if (winner === 0) {
            setStatus('Draw!');
        } else if (winner === 1) {
            setStatus('Winner: X');
        } else if (winner === -1) {
            setStatus('Winner: O');
        }
    }

    return (
        <div>
            <div className="status">{status}</div>
            <div className="row">
                {renderSquare(0,0)}
                {renderSquare(0,1)}
                {renderSquare(0,2)}
            </div>
            <div className="row">
                {renderSquare(1,0)}
                {renderSquare(1,1)}
                {renderSquare(1,2)}
            </div>
            <div className="row">
                {renderSquare(2,0)}
                {renderSquare(2,1)}
                {renderSquare(2,2)}
            </div>
            <button className='square reset-button' onClick={resetGame}>Reset</button>
        </div>
    );
};

export default TicTacToe;
